### PR TITLE
Add Table ID routing logic and validation For MultiNic feature.

### PR DESCRIPTION
**Pull request by halimsam**
This change introduces validation and routing configuration for MultiNic. MultiNic is enabled by node label `lustre.csi.storage.gke.io/multi-rail=true`.  If enabled,  we add all the NICs as lnet network parameters during kmod install. When a lustre instance is created, attached to a pod, and the pod is scheduled to a node with MultiNic enabled: we configure routes for the NIC and Lustre instance on the Table ID.  Each available NIC (except for default nic0) will be paired to a unique Table ID and routes will be configured to all Nic's Table ID for each Lustre instance. You can also create the nodepool node label directly, eliminating the need to update them after node creation.

This diagram below shows the routing table when NodeStageVolume is called with MultiNic logic.

Diagram:
image

Successful lustre CSI setup:
image

successful routing logs setup for nics during mounting (NodeStageVolume):
image

Pod running (using dynamic-pvc.yaml and dynamic-pod.yaml):
image

**For testing:**
I created my own network with lustre setup, and added a non-overlap subnet for the additional nic (multinic-subnet)

1. I created a cluster/nodepool with multiple VPC/subnet to have multi nics:
```
gcloud container clusters create single-vpc \ --location=us-central1-c \ --num-nodes=1 \ --network=samhalim-network \ --cluster-version=1.34.1-gke.2541000 --enable-kernel-module-signature-enforcement
```

2. Create custom network with subnet:
```
gcloud container clusters create single-vpc --location=us-central1-c --num-nodes=1 --network=samhalim-network --cluster-version=1.34.1-gke.2541000 --enable-kernel-module-signature-enforcement
```

3. Create nodepool with multi nic
```
gcloud container node-pools create single-vpc --cluster=single-vpc --num-nodes=1 --zone=us-central1-c --additional-node-network network=samhalim-network,subnetwork=multinic-subnet --enable-kernel-module-signature-enforcement
```

4. Add multi-rail node label to node with multi nics (eth0, eth1):
```
kubectl label node <NODE_NAME> lustre.csi.storage.gke.io/multi-rail=true
```

### CODE REVIEW

File: pkg/network/network.go [hunk 1]
```diff
@@ -17,15 +17,22 @@ limitations under the License.
...
 
+	"github.com/GoogleCloudPlatform/lustre-csi-driver/pkg/k8sclient"
 	"github.com/vishvananda/netlink"
 	"k8s.io/klog/v2"
 )
 
 const (
 	multiRailLabel = "lustre.csi.storage.gke.io/multi-rail"
+	// lustreInstanceIP is the IP of the Lustre instance passed in through volumeContext.
+	lustreInstanceIP = "10.170.0.0/16"
```
Comment (tyuchn) [inline pkg/network/network.go]: Instance IP is dynamic, not constant. It can only be retrieved from volumeContext during NodeStageVolume.

File: pkg/kmod_installer/kmod_installer.go [hunk 2]
```diff
@@ -165,5 +166,17 @@ func InstallLustreKmod(ctx context.Context, nodeID string, enableLegacyPort bool
 
 	klog.Infof("Cos-dkms output:\n%s\n", string(output))
 
+	if isMultiNic {
+		tableID := initialRouteTableID
+		for _, nicName := range nics {
+			klog.Info(nicName)
```
[the remark this answers is not in the document]
Comment (halimsam) [reply pkg/kmod_installer/kmod_installer.go]: Removed.

File: pkg/network/network.go [hunk 3]
```diff
@@ -46,13 +53,166 @@ func GetGvnicNames() ([]string, error) {
...
+
+	if gateway == nil {
+		return fmt.Errorf("could not find gateway for %s", nicName)
+	}
+	klog.Infof("Found gateway %s for %s", gateway.String(), nicName)
+
+	// Define and add the route
+	// This is the IP of the Lustre instance which is passed in through volumeContext.
+	instanceIPAddr += "/32"
+	_, dst, _ := net.ParseCIDR(instanceIPAddr) // MGS IP
```
Comment (tyuchn) [inline pkg/network/network.go]: Remove this unused comment and below

**Review by tyuchn — COMMENTED** [review]
LGTM with several comments. I'm also a bit concerned about how we can verify the multi-NIC related functions offline, like through unit tests, as end-to-end testing in production is expensive and might miss some corner cases.

File: cmd/csi_driver/main.go [hunk 4]
```diff
@@ -68,18 +69,42 @@ func main() {
...
 	}
+	// additional NICs are any additional NICs that are not default (eth0).
+	// These NICs will additional setup for multi nic feature.
+	additionalNics := []string{}
+	for _, nic := range nics {
+		if nic != "eth0" {
+			additionalNics = append(additionalNics, nic)
+		}
+	}
+	klog.V(4).Infof("additional nic(s) %v on Node %v", additionalNics, nodeID)
```
[the remark this answers is not in the document]
Comment (halimsam) [reply cmd/csi_driver/main.go]: Done.

File: pkg/csi_driver/node.go [hunk 5]
```diff
@@ -158,6 +160,28 @@ func (s *nodeServer) NodeStageVolume(_ context.Context, req *csi.NodeStageVolume
 		return nil, status.Errorf(codes.Internal, "Could not mount %q at %q on node %s: %v", source, target, nodeName, err)
 	}
 
+	if s.driver.config.IsMultiNic {
+		klog.Info("Multi Nic feature is enabled and will configure route for Lustre instance.")
```
[the remark this answers is not in the document]
Comment (halimsam) [reply pkg/csi_driver/node.go]: Done.

File: pkg/network/network.go [hunk 6]
```diff
@@ -46,13 +53,166 @@ func GetGvnicNames() ([]string, error) {
...
+
+	if err := configureRoutesForNic(nicName, nicIPAddr, instanceIP, tableID); err != nil {
+		return fmt.Errorf("failed to configure routes for %s: %w", nicName, err)
+	}
+
 	return nil
 }
 
-func IsMultiRailEnabled(nodeID string) (bool, error) {
+func GetNicIPAddr(nicName string) (net.IP, error) {
```
[the remark this answers is not in the document]
Comment (halimsam) [reply pkg/network/network.go]: Done. I should have done this with all of the exported functions related to this PR.

**Review by tyuchn — APPROVED** [review]
Approved as we will explore incorporating more unit tests for multi-NIC related functions in a follow-up PR discussed
